relay-server/src/actors/project_buffer.rs

```diff
@@ -0,0 +1,181 @@
+use std::collections::{BTreeMap, BTreeSet};
+
+use tokio::sync::mpsc;
+
+use relay_common::ProjectKey;
+use relay_system::{FromMessage, Interface, Service};
+
+use crate::envelope::Envelope;
+use crate::utils::EnvelopeContext;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct QueueKey {
+    pub own_key: ProjectKey,
+    pub sampling_key: ProjectKey,
+}
+
+impl QueueKey {
+    pub fn new(own_key: ProjectKey, sampling_key: ProjectKey) -> Self {
+        Self {
+            own_key,
+            sampling_key,
+        }
+    }
+}
+
+/// Adds the envelope and the envelope context to the internal buffer.
+#[derive(Debug)]
+pub struct Enqueue {
+    key: QueueKey,
+    value: (Box<Envelope>, EnvelopeContext),
+}
+
+impl Enqueue {
+    pub fn new(key: QueueKey, value: (Box<Envelope>, EnvelopeContext)) -> Self {
+        Self { key, value }
+    }
+}
+
+/// Removes messages from the internal buffer and streams them to the sender.
+#[derive(Debug)]
+pub struct DequeueMany {
+    keys: Vec<QueueKey>,
+    sender: mpsc::UnboundedSender<(Box<Envelope>, EnvelopeContext)>,
+}
+
+impl DequeueMany {
+    pub fn new(
+        keys: Vec<QueueKey>,
+        sender: mpsc::UnboundedSender<(Box<Envelope>, EnvelopeContext)>,
+    ) -> Self {
+        Self { keys, sender }
+    }
+}
+
+/// Removes the provided keys from the internal buffer.
+#[derive(Debug)]
+pub struct RemoveMany {
+    project_key: ProjectKey,
+    keys: BTreeSet<QueueKey>,
+}
+
+impl RemoveMany {
+    pub fn new(project_key: ProjectKey, keys: BTreeSet<QueueKey>) -> Self {
+        Self { project_key, keys }
+    }
+}
+
+/// The Interface for [`BufferService`] service.
```

Comment:
_nit_: We could describe a bit better here what this interface is for (and how to use it, if necessary).